### PR TITLE
Pull request to merge Flowchart page with CalcUS 2.0

### PR DESCRIPTION
**Pull request by div-yam**

### CODE REVIEW

File: frontend/management/commands/dummy_commands.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
+from frontend.models import (
```
Comment (RaphaelRobidas) [inline frontend/management/commands/dummy_commands.py:1]: Please untrack this file for the PR: `git rm --cached frontend/management/commands/dummy_commands.py`

File: frontend/models.py [hunk 2]
```diff
@@ -18,8 +18,10 @@
 """
 
 
+from tkinter import CASCADE
```
Comment (RaphaelRobidas) [inline frontend/models.py:21]: This line is not needed (you were looking for models.CASCADE)

File: frontend/models.py [hunk 3]
```diff
@@ -1196,11 +1311,21 @@ def get_mol(self):
...
+            old_unseen = self.flowchart_order.new_status
+
+        elif hasattr(self, "order"):
+            old_status = self.order.status
+            old_unseen = self.order.new_status
 
         super(Calculation, self).save(*args, **kwargs)
-        self.order.update_unseen(old_status, old_unseen)
+
+        if hasattr(self, "flowchart_order") and self.flowchart_order is not None:
```
Comment (RaphaelRobidas) [inline frontend/models.py:1324]: You do not need to check if the object has these attributes, they will always be created and equal to `None` if unset. You could also verify if the `order` if is null as second case, and raise an exception if both `flowchart_order` and `order` are null.

File: frontend/tasks.py [hunk 4]
```diff
@@ -20,6 +20,8 @@
 
 from __future__ import absolute_import, unicode_literals
 
+# from asyncio.windows_events import NULL
```
Comment (RaphaelRobidas) [inline frontend/tasks.py:23]: Lines to clean up

File: frontend/tasks.py [hunk 5]
```diff
@@ -3272,15 +3274,41 @@ def get_rel_energy(s):
 
 
 @app.task(base=AbortableTask)
-def dispatcher(order_id):
-    order = CalculationOrder.objects.get(pk=order_id)
-    ensemble = order.ensemble
+def dispatcher_flowchart():
```
Comment (RaphaelRobidas) [inline frontend/tasks.py:3277]: For debugging only, I suppose? To remove if it is no longer necessary

File: frontend/tasks.py [hunk 6]
```diff
@@ -3272,15 +3274,41 @@ def get_rel_energy(s):
...
+@app.task(base=AbortableTask)
+def dispatcher(order_id, drawing=None, is_flowchart=False, flowchartStepObjectId=None):
+    stepFlowchart = None
+    should_create_ensemble = True
+    if is_flowchart:
+        order = FlowchartOrder.objects.get(pk=order_id)
+        flowchartStepObject = Step.objects.get(pk=flowchartStepObjectId)
+        if flowchartStepObject.parentId.step is not None:
+            if flowchartStepObject.parentId.step.creates_ensemble is True:
+                if len(flowchartStepObject.calculation_set.all()) != 0:
```
Comment (RaphaelRobidas) [inline frontend/tasks.py:3290]: You can use `flowchartStepObject.calculation_set.count()` to get the number of results (it's cleaner and perhaps marginally more efficient).

File: frontend/tasks.py [hunk 7]
```diff
@@ -3272,15 +3274,41 @@ def get_rel_energy(s):
...
+                    step_ensemble = flowchartStepObject.calculation_set.all()[
```
Comment (RaphaelRobidas) [inline frontend/tasks.py:3291]: You can use `flowchartStepObject.calculation_set.first()`. The case where the number of results is greater than one should be handled at some point in the future, perhaps throw an exception for now.

**Review by RaphaelRobidas — CHANGES_REQUESTED** [review]
Great work! There are just a couple minor changes for you to make, otherwise the code looks good.

File: frontend/templates/frontend/flowchart.html [hunk 8]
```diff
@@ -0,0 +1,902 @@
...
+	var para_calc_id
+	function verify_form_flowchart() {
+		data = $("#calcform").serializeArray()
+		form = $("#calcform");
+		formData = new FormData(form[0])
+		calc_para_map.set(parseInt(para_calc_id), formData)
+		data.push({"name": "constraint_num", "value": constraint_num});
+
+		{% if ensemble or structures %}
+		data.push({"name": "starting_ensemble", "value": {{ ensemble.id }}});
```
Comment (RaphaelRobidas) [inline frontend/templates/frontend/flowchart.html:672]: The ids are not hashes, they need to be treated as strings within quotation marks

File: frontend/templates/frontend/flowchart.html [hunk 9]
```diff
@@ -0,0 +1,902 @@
...
+		{% endif %}
+
+		{% if structures %}
+		data.push({"name": "starting_structs", "value": "{{ structures }}"});
+		{% endif %}
+
+		{% if calc %}
+		data.push({"name": "starting_calc", "value": {{ calc.id }}});
```
Comment (RaphaelRobidas) [inline frontend/templates/frontend/flowchart.html:680]: See last comment

File: frontend/management/commands/dummy_commands.py (hunk 1 above)
Comment (div-yam) [reply frontend/management/commands/dummy_commands.py:1]: Done!

File: frontend/models.py (hunk 2 above)
Comment (div-yam) [reply frontend/models.py]: Done!

File: frontend/models.py (hunk 3 above)
Comment (div-yam) [reply frontend/models.py]: Updated!

File: frontend/tasks.py (hunk 4 above)
Comment (div-yam) [reply frontend/tasks.py]: Done!

File: frontend/tasks.py (hunk 5 above)
Comment (div-yam) [reply frontend/tasks.py]: Yes, it was for debugging only. Removed.

File: frontend/tasks.py (hunk 6 above)
Comment (div-yam) [reply frontend/tasks.py]: Done!

File: frontend/tasks.py (hunk 7 above)
Comment (div-yam) [reply frontend/tasks.py]: Done and exception added.

File: frontend/templates/frontend/flowchart.html (hunk 8 above)
Comment (div-yam) [reply frontend/templates/frontend/flowchart.html]: Done!

File: frontend/templates/frontend/flowchart.html (hunk 9 above)
Comment (div-yam) [reply frontend/templates/frontend/flowchart.html]: Done!

**Pull request merged**